os-eap-sso/added/keycloak.sh

```diff
@@ -464,5 +474,97 @@ function get_application_routes {
+    if [ "$secure" = true ]; then
+      local secureroute=$(build_route_url "secure-${route}")
+      # removing default port since keycloak only matches the exactly URL
+      if [[ "${secureroute},," = *"443"* ]]; then
+        echo $secureroute | awk -F ":" '{ print substr($0,0,length($0)-length($NF)-1) }'
+      else
+        echo $secureroute
+      fi
+    else
+      build_route_url "${route}"
+    fi
+}
 
+# Queries the Route from the Kubernetes API
+# ${1} - route name
+query_route() {
+    local routeName=${1}
+    # only execute the following lines if this container is running on OpenShift
+    if [ -e /var/run/secrets/kubernetes.io/serviceaccount/token ]; then
+        local namespace=$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace)
+        local token=$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)
+        local response=$(curl -s -w "%{http_code}" --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt \
+            -H "Authorization: Bearer $token" \
+            -H 'Accept: application/json' \
+            https://${KUBERNETES_SERVICE_HOST:-kubernetes.default.svc}:${KUBERNETES_SERVICE_PORT:-443}/apis/route.openshift.io/v1/namespaces/${namespace}/routes/${routeName})
+        echo ${response}
+    fi
 }
+
+# Queries the Route host from the Kubernetes API
+# ${1} - route name
+# ${2} - default host
+query_route_host() {
+    local routeName=${1}
+    local host=${2}
+    if [ "${routeName}" != "" ]; then
+        local response=$(query_route "${routeName}")
+        if [ "${response: -3}" = "200" ]; then
+            # parse the json response to get the route host
+            host=$(echo ${response::- 3} | python -c 'import json,sys;obj=json.load(sys.stdin);print obj["spec"]["host"]')
+        else
+            log_warning "Fail to query the Route using the Kubernetes API, the Service Account might not have the necessary privileges; defaulting to host [${host}]."
+            if [ ! -z "${response}" ]; then
+                log_warning "Response message: ${response::- 3} - HTTP Status code: ${response: -3}"
+            fi
+        fi
+    fi
+    echo "${host}"
+}
+
+# Builds a simple URL
+# ${1} - protocol (default is http)
+# ${2} - host (default is $HOSTNAME if possible, or localhost)
+# ${3} - port (default is empty)
+# ${4} - path (default is empty)
+build_simple_url() {
+    local protocol=${1:-http}
+    local host=${2:-${HOSTNAME:-localhost}}
+    local port=${3}
+    local path=${4}
+    if [ "${port}" != "" ]; then
+        port=":${port}"
+    fi
+    echo "${protocol}://${host}${port}${path}"
+}
+
+# Builds a Route URL by querying the Kubernetes API
+# ${1} - route name
+# ${2} - default protocol
+# ${3} - default host
+# ${4} - default port
+# ${5} - path
+build_route_url() {
+    local routeName=${1}
+    local protocol=${2}
+    local host=${3}
+    local port=${4}
+    local path=${5}
+    if [ "${routeName}" != "" ]; then
+        if [[ "${routeName},," = *"secure"* ]]; then
```

Comment:
This makes the assumption that route names starting with "secure" is how we determine this. I know this is the way the code is that this script borrows from (jboss-kie-modules), however we have control over the route names there, whereas here (in a more generic usage), we do not.

Reply:
Is it true that if the route is secure, then by definition the protocol is https? I guess I could see some configurations where the route is internal only and not https. I do agree with Davie on the naming thing, though it is certainly the OS default naming.

Reply:
If it is the OS default naming, then my worry goes away.

Reply:
I think that OS will suffix it with -secure rather than prefix it. I'll have a look in OS to check.

Reply:
Using an example from the default registry, I see the api return: 
`{
	"kind": "RouteList",
	"apiVersion": "route.openshift.io/v1",
	"metadata": {
		"selfLink": "/apis/route.openshift.io/v1/namespaces/default/routes/",
		"resourceVersion": "409231"
	},
	"items": [{
		"metadata": {
			"name": "docker-registry",
			"namespace": "default",
			"selfLink": "/apis/route.openshift.io/v1/namespaces/default/routes/docker-registry",
			"uid": "35364faa-6fbc-11e9-8fa4-c85b76aab2ff",
			"resourceVersion": "2359",
			"creationTimestamp": "2019-05-06T05:02:54Z",
			"labels": {
				"docker-registry": "default"
			},
			"annotations": {
				"openshift.io/host.generated": "true"
			}
		},
		"spec": {
			"host": "docker-registry-default.127.0.0.1.nip.io",
			"to": {
				"kind": "Service",
				"name": "docker-registry",
				"weight": 100
			},
			"port": {
				"targetPort": "5000-tcp"
			},
			"tls": {
				"termination": "edge"
			},
			"wildcardPolicy": "None"
		},
		"status": {
			"ingress": [{
				"host": "docker-registry-default.127.0.0.1.nip.io",
				"routerName": "router",
				"conditions": [{
					"type": "Admitted",
					"status": "True",
					"lastTransitionTime": "2019-05-06T05:02:54Z"
				}],
				"wildcardPolicy": "None"
			}]
		}
	}]
}`

This route has is deemed to be 'secure' in the OS console, the only thing present in the json I see is the tls bit.

Reply:
Same route with secure disabled in console:

<pre>
{
	"kind": "Route",
	"apiVersion": "route.openshift.io/v1",
	"metadata": {
		"name": "docker-registry",
		"namespace": "default",
		"selfLink": "/apis/route.openshift.io/v1/namespaces/default/routes/docker-registry",
		"uid": "35364faa-6fbc-11e9-8fa4-c85b76aab2ff",
		"resourceVersion": "410725",
		"creationTimestamp": "2019-05-06T05:02:54Z",
		"labels": {
			"docker-registry": "default"
		},
		"annotations": {
			"openshift.io/host.generated": "true"
		}
	},
	"spec": {
		"host": "docker-registry-default.127.0.0.1.nip.io",
		"to": {
			"kind": "Service",
			"name": "docker-registry",
			"weight": 100
		},
		"port": {
			"targetPort": "5000-tcp"
		},
		"wildcardPolicy": "None"
	},
	"status": {
		"ingress": [{
			"host": "docker-registry-default.127.0.0.1.nip.io",
			"routerName": "router",
			"conditions": [{
				"type": "Admitted",
				"status": "True",
				"lastTransitionTime": "2019-05-06T05:02:54Z"
			}],
			"wildcardPolicy": "None"
		}]
	}
}
</pre>

Reply:
I'll take a different path here. I'm going to query for the service that has the same name as the pod's. With the service we can get every route related and then use it as URLs. Makes much more sense.

Unfortunately, "secure" is a name that we had in BA images, I've tested the script with other images and only the non-secure one was retrieved. :(

Reply:
Yes, the only difference for the "secure" ones is the tls property. :(